### PR TITLE
Havardpe/more efficient single outer reduce

### PR DESCRIPTION
**Pull request by havardpe**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

@arnej27959 please review

### CODE REVIEW

File: eval/src/vespa/eval/eval/aggr.h [hunk 1]
```diff
@@ -156,17 +186,21 @@ template <typename T> class Median {
...
     constexpr Min() : _min{std::numeric_limits<T>::infinity()} {}
     constexpr Min(T value) : _min{value} {}
     constexpr void sample(T value) { _min = std::min(_min, value); }
     constexpr void merge(const Min &rhs) { _min = std::min(_min, rhs._min); }
     constexpr T result() const { return _min; }
+    static constexpr Aggr enum_type() { return Aggr::MIN; }
+    static T combine(T a, T b) { return std::min(a,b); }
 };
 
 } // namespave vespalib::eval::aggr
```
Comment (arnej27959) [inline eval/src/vespa/eval/eval/aggr.h:206]: namespave -> namespace

File: eval/src/vespa/eval/eval/aggr.cpp [hunk 2]
```diff
@@ -17,6 +17,7 @@ struct Wrapper : Aggregator {
     virtual void first(double value) final override { aggr = T{value}; }
     virtual void next(double value) final override { aggr.sample(value); }
     virtual double result() const final override { return aggr.result(); }
+    virtual Aggr enum_type() const final override { return T::enum_type(); }
```
Comment (arnej27959) [inline eval/src/vespa/eval/eval/aggr.cpp:20]: while "enum_type" works well, I think I would have chosen to call it "enum_value" or "aggr_enum" instead.

File: eval/src/vespa/eval/eval/aggr.h [hunk 3]
```diff
@@ -76,56 +93,69 @@ template <typename T> class Avg {
...
     T _sum;
 public:
+    using value_type = T;
     constexpr Sum() : _sum{0} {}
     constexpr Sum(T value) : _sum{value} {}
     constexpr void sample(T value) { _sum += value; }
     constexpr void merge(const Sum &rhs) { _sum += rhs._sum; }
     constexpr T result() const { return _sum; }
+    static constexpr Aggr enum_type() { return Aggr::SUM; }
+    static T combine(T a, T b) { return (a += b); }
```
Comment (arnej27959) [inline eval/src/vespa/eval/eval/aggr.h:137]: is it useful to have `a += b` here?  As it's used from
`*dst = combine(*dst, *src++);`
it gives me some doubt.   I guess you also considered formulation with
`static void combine(T& a, T b) { a += b; }`
and so on.

**Review by arnej27959 — APPROVED** [review]
only minor cosmetic suggestions from me, look very nice

File: eval/src/vespa/eval/eval/aggr.h (hunk 3 above)
Comment (havardpe) [reply eval/src/vespa/eval/eval/aggr.h:137]: yes, was a relic from the 'past'

**Comment by havardpe** [comment]
@arnej27959 PTAL
I have fixed all your comments and also added constexpr to the combine functions

**Review by arnej27959 — APPROVED** [review]
👍

**Pull request merged**